crates/fluvio/src/producer/mod.rs

```diff
@@ -29,355 +35,301 @@ pub struct TopicProducer {
         let partitioner = Box::new(SiphashRoundRobinPartitioner::new());
+
+        let (to_dispatcher, from_dispatcher) = flume::bounded(0);
+        let (status_sender, status_receiver) = tokio::sync::broadcast::channel(100);
+        let dispatcher =
+            Dispatcher::new(pool.clone(), status_sender.clone(), from_dispatcher, config);
+        let dispatcher_shutdown = dispatcher.start();
+        let inner = Arc::new(ProducerInner {
+            dispatcher: to_dispatcher,
+            dispatcher_shutdown,
+            uid_generator: RecordUidGenerator::new(),
+            _status_sender: status_sender,
+            status_receiver: Mutex::new(status_receiver),
+        });
+
         Self {
             topic,
             pool,
             partitioner,
+            inner,
         }
     }
 
     /// Sends a key/value record to this producer's Topic.
     ///
     /// The partition that the record will be sent to is derived from the Key.
     ///
+    /// The `send` function does not wait for records to be committed to the SPU,
+    /// it simply buffers them. In order to make sure that your records have all
+    /// been actually sent, be sure to call [`TopicProducer::flush`] after sending
+    /// all the records you have.
+    ///
     /// # Example
     ///
     /// ```
     /// # use fluvio::{TopicProducer, FluvioError};
     /// # async fn example(producer: &TopicProducer) -> Result<(), FluvioError> {
     /// producer.send("Key", "Value").await?;
+    /// producer.flush().await?;
     /// # Ok(())
     /// # }
     /// ```
     #[instrument(
         skip(self, key, value),
         fields(topic = %self.topic),
     )]
-    pub async fn send<K, V>(&self, key: K, value: V) -> Result<(), FluvioError>
+    pub async fn send<K, V>(&self, key: K, value: V) -> Result<RecordUid>
     where
         K: Into<RecordKey>,
         V: Into<RecordData>,
     {
-        let record_key = key.into();
-        let record_value = value.into();
-        self.send_all(Some((record_key, record_value))).await?;
-        Ok(())
+        let key = key.into();
+        let value = value.into();
+        let record = Record::from((key, value));
+
+        {
+            // Check if there have been any failures
+            let mut receiver = self.inner.status_receiver.lock().expect("Mutex poisoned");
+            while let Ok(status) = (*receiver).try_recv() {
+                match status {
+                    BatchStatus::Failure(failed) => {
+                        return Err(ProducerError::BatchFailed(failed).into());
+                    }
+                    BatchStatus::InternalError(e) => {
+                        return Err(FluvioError::InternalProducerError(e));
+                    }
+                    // No action to take on success
+                    _ => {}
+                }
+            }
+        }
+
+        let topic_spec = self.pool.lookup_topic(&self.topic).await?.spec;
```

Comment:
this should been already computed. no need to do again for every send call

Reply:
Isn't the point of the MetadataStores that the metadata could change? This shouldn't end up calculating anything, just look up the spec in the cache